File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/DeletePostController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.dto.RecordDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.repository.HistoryRepository
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.RecordSerializer
import org.springframework.http.HttpStatus
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class DeletePostController(
    private val recordRepository: RecordRepository,
    private val historyRepository: HistoryRepository,
    private val recordSerializer: RecordSerializer
) {
    data class Params(
        val id: Int,
        val recordOptions: RecordSerializer.Options?
    )
    data class Result(
        val record: RecordDTO?
    )

    @PostMapping("/v3/DeletePost")
    @Transactional
    fun handle(@RequestBody params: Params, @CurrentUser currentUser: User): Result {
        val history = historyRepository.findById(params.id).orElseThrow { ApiException.notFound() }
        if (currentUser.id != history.user.id) {
            throw ApiException.permissionDenied()
        }
        val record = history.record
        if (historyRepository.countByRecord(record) == 1) {
            throw ApiException("등록된 작품마다 최소 1개의 기록이 필요합니다.", HttpStatus.UNPROCESSABLE_ENTITY)
        }
        historyRepository.delete(history)
        val latestHistory = historyRepository.findFirstByRecordOrderByIdDesc(record)!!
        record.status = latestHistory.status
        record.status_type = latestHistory.status_type
        recordRepository.save(record)
        return Result(
            record = params.recordOptions?.let { recordSerializer.serialize(record, params.recordOptions) }
        )
    }
}

File: backend-ts/src/controllers/delete_post.controller.ts
import { Body, Controller, HttpCode, HttpStatus, Post } from "@nestjs/common";
import { RecordDTO } from 'shared/types';
import { CurrentUser } from "src/auth/decorators";
import { History } from "src/entities/history.entity";
import { User } from "src/entities/user.entity";
import { RecordSerializer, RecordSerializerOptions } from "src/serializers/record.serializer";
import { Connection } from "typeorm";
import { ApiException } from "./exceptions";

type Params = {
  id: number;
  recordOptions?: RecordSerializerOptions;
}

type Result = {
  record: RecordDTO | null;
}

@Controller('/api/v4/DeletePost')
export class DeletePostController {
  constructor(
    private connection: Connection,
    private recordSerializer: RecordSerializer,
  ) {}

  @Post()
  @HttpCode(200)
  async handle(
    @Body() params: Params,
    @CurrentUser() currentUser: User,
  ): Promise<Result> {
    const updatedRecord = await this.connection.transaction(async em => {
      const history = await em.findOne(History, params.id, {relations: ['record']})
      if (!history)
        throw ApiException.notFound()
      if (currentUser.id !== history.user_id)
        throw ApiException.permissionDenied()
      const record = history.record
      if (await em.count(History, { where: { record } }) === 1)
        throw new ApiException('등록된 작품마다 최소 1개의 기록이 필요합니다.', HttpStatus.UNPROCESSABLE_ENTITY)
      await em.remove(history)

      const latestHistory = (await em.find(History, {
        where: {record},
        order: {id: 'DESC'},
        take: 1,
      }))[0]
      record.status = latestHistory.status
      record.status_type = latestHistory.status_type
      await em.save(record)

      return record
    })

    return {
      record: params.recordOptions ? await this.recordSerializer.serialize(updatedRecord, params.recordOptions) : null,
    }
  }
}
